pgjdbc/src/test/java/org/postgresql/replication/LogicalReplicationTest.java

```diff
@@ -0,0 +1,637 @@
+
+    ByteBuffer result = stream.readPending();
+
+    assertThat("Read pending message allow without lock on socket read message, "
+            + "and if message absent return null. In current test we start replication from last LSN on server, "
+            + "so changes absent on server and readPending message will always lead to null ByteBuffer",
+        result, equalTo(null)
+    );
+  }
+
+  @Test(timeout = 3000)
+  public void testReadPreviousChangesWithoutBlock() throws Exception {
+    PGConnection pgConnection = (PGConnection) replConnection;
+
+    LogSequenceNumber startLSN = getCurrentLSN();
+
+    Statement st = sqlConnection.createStatement();
+    st.execute("insert into test_logic_table(name) values('previous changes')");
+    st.close();
+
+    PGReplicationStream stream =
+        pgConnection
+            .replicationStream()
+            .logical()
+            .withSlotName(SLOT_NAME)
+            .withStartPosition(startLSN)
+            .withSlotOption("include-xids", false)
+            .withSlotOption("skip-empty-xacts", true)
+            .start();
+
+    String received = group(receiveMessageWithoutBlock(stream, 3));
+
+    String wait = group(Arrays.asList(
+        "BEGIN",
+        "table public.test_logic_table: INSERT: pk[integer]:1 name[character varying]:'previous changes'",
+        "COMMIT"
+    ));
+
+    assertThat(
+        "Messages from stream can be read by readPending method for avoid long block on Socket, "
+            + "in current test we wait that behavior will be same as for read message with block",
+        received, equalTo(wait)
+    );
+  }
+
+  @Test(timeout = 3000)
+  public void testReadActualChangesWithoutBlock() throws Exception {
+    PGConnection pgConnection = (PGConnection) replConnection;
+
+    PGReplicationStream stream =
+        pgConnection
+            .replicationStream()
+            .logical()
+            .withSlotName(SLOT_NAME)
+            .withStartPosition(getCurrentLSN())
+            .withSlotOption("include-xids", false)
+            .withSlotOption("skip-empty-xacts", true)
+            .start();
+
+    Statement st = sqlConnection.createStatement();
+    st.execute("insert into test_logic_table(name) values('actual changes')");
+    st.close();
+
+    String received = group(receiveMessageWithoutBlock(stream, 3));
+
+    String wait = group(Arrays.asList(
+        "BEGIN",
+        "table public.test_logic_table: INSERT: pk[integer]:1 name[character varying]:'actual changes'",
+        "COMMIT"
+    ));
+
+    assertThat(
+        "Messages from stream can be read by readPending method for avoid long block on Socket, "
+            + "in current test we wait that behavior will be same as for read message with block",
+        received, equalTo(wait)
+    );
+  }
+
+  @Test(timeout = 30000 /* backend keep alive 10s * 3 */)
+  public void testAvoidTimeoutDisconnectWithDefaultStatusInterval() throws Exception {
+    exception.expect(TestTimedOutException.class);
```

Comment:
Is "keep alive" configurable? Can we make it 3-4 seconds?
It sounds sad to have a test that takes 30 seconds.